feat: Add read-only mode for terminal sessions

ChetanXpro opened the pull request:
- Add support for read-only viewer access via --enable-readers flag
- Implement write permissions check for terminal operations
- New users default to read-only mode if no write password is provided.
- Add visual indicator for read-only mode in UI
- Disable terminal editing and controls for read-only users

Fixes #72 

Demo:
https://github.com/user-attachments/assets/95434d88-5895-4db2-ad2e-577bbb061616

ekzhang commented:
Woohoo! The video you sent looks great. Let me check the code now, there's also a couple minor changes we could make to the interface.

ekzhang reviewed (commented):
This is fantastic! Thanks for implementing this. I'm super excited to have this feature out. I think we could add tests + figure out how to unify the same `Encrypt` struct for the two kinds of passwords (auth key and write key) and we'd be set.



<img width="250" alt="image" src="https://github.com/user-attachments/assets/6ab801c0-4220-46c2-ae12-ed15289127ff" />

As for the display on the frontend, I like what you did with stopping the events from going off. The server error handling also looks robust. In this display though, I think we might want to make the "Read Only" badge smaller, maybe move it to the right of the status line. And in the `NameList` component, move the read-only users below the read-write users, and gray out their names with lower opacity? What do you think?

Something like this:

- Current (if not `--enable-readers`): https://play.tailwindcss.com/2sBMyEFomE
  <img width="306" alt="image" src="https://github.com/user-attachments/assets/db015109-393d-45be-ad8e-99dff91ec9c8" />
- New (if having `--enable-readers`): https://play.tailwindcss.com/X66EJc5hdD
  <img width="350" alt="image" src="https://github.com/user-attachments/assets/9fb41049-0f22-416c-8d24-be5724b70f6a" />



Having read-only users not be able to move terminals might be a point of contention, but I think it makes sense personally! This way you can share the link more widely with a group without having to worry about people potentially resizing or moving stuff. And in practice it's probably not too bad, since the editor will usually position the terminal how they would like.

Happy holidays! :)

ChetanXpro commented on Cargo.lock:
```diff
@@ -440,6 +440,12 @@ dependencies = [
 ]
 
+[[package]]
+name = "constant_time_eq"
```
Yup, that makes sense. I will use `subtle`. Yeah, 
I noticed we are not using encrypted_zeros with constant-time, but I thought maybe it was intentional.

ChetanXpro commented on crates/sshx-server/src/grpc.rs:
```diff
@@ -50,12 +50,20 @@ impl SshxService for GrpcServer {
         info!(%name, "creating new session");
+
+        let write_password = if request.enable_readers {
+            Some(rand_alphanumeric(8))
```
I agree, indeed it's very sketchy, but yeah, I can use Encrypt::zeros for write_password as well. It shouldn't take much time

ChetanXpro commented:
Thanks for the feedback! I’ll add the tests and unify the Encrypt struct. I agree with your suggestions for the frontend changes, and I’ll make those adjustments

I’ll fix everything and update the PR. Happy holidays!

ChetanXpro commented:
fixed most of the issues. tests and the Encrypt struct updates are remaining - working on those now.

now we are now creating the write password in the controller itself and only saving the write password encryption zeros in metadata and snapshot.

ekzhang commented on crates/sshx-server/src/web/socket.rs:
```diff
@@ -95,15 +96,45 @@ async fn handle_socket(socket: &mut WebSocket, session: Arc<Session>) -> Result<
+        Some(WsClient::Authenticate(bytes, write_password_bytes)) => {
+            // `ct_eq` returns a `Choice`, and `unwrap_u8()` converts it to 1 (equal) or 0
+            // (not equal).
+            if bytes.ct_eq(metadata.encrypted_zeros.as_ref()).unwrap_u8() != 1 {
```
<img width="1040" alt="image" src="https://github.com/user-attachments/assets/7d2b9de4-b98d-4046-a8a7-9695c45b434b" />

nit: shouldn't this do `into()` to use the From impl? I usually wouldn't bring it up except it seems like they bolded it for emphasis in the docs :)

ekzhang commented on crates/sshx/src/controller.rs:
```diff
@@ -46,25 +47,51 @@ pub struct Controller {
             task::spawn_blocking(move || Encrypt::new(&encryption_key))
         };
+
+        let kdf_write_password_task = {
```
small change: Can we move the `if enable_readers` branch up one level so we don't need to do argon2 construction if it's not used? It's moved into a separate thread since it might take some time. (couple hundred millis)

ekzhang commented on crates/sshx/src/main.rs:
```diff
@@ -26,27 +26,51 @@ struct Args {
+            r#"
+      {sshx} {version}
+    
+      {arr}  Read-only link:  {link_v}
+      {arr}  Writable  link:  {link_e}
+      {arr}  Shell:           {shell_v}
+    "#,
```
```suggestion
      {arr}  Read-only link: {link_v}
      {arr}  Writable link:  {link_e}
      {arr}  Shell:          {shell_v}
"#,
```

ekzhang commented on crates/sshx/src/main.rs:
```diff
@@ -26,27 +26,51 @@ struct Args {
+    
+      {arr}  Link:  {link_v}
+      {arr}  Shell: {shell_v}
+    "#,
```
```suggestion
"#,
```

ekzhang commented on src/lib/Session.svelte:
```diff
@@ -127,10 +129,19 @@
     encrypt = await Encrypt.new(key);
     const encryptedZeros = await encrypt.zeros();
 
+    write_password_encrypt = writePassword
```
You can just save this as a temporary variable, no need to put it as a prop at the top level of the svelte component!

Also `writePasswordEncrypt` naming

ekzhang commented on src/lib/Session.svelte:
```diff
@@ -445,7 +467,33 @@
+          <div
+            class="bg-yellow-900 text-yellow-200 px-1 py-0.5 rounded ml-3 inline-flex items-center gap-1"
+          >
+            <svg
+              class="w-3.5 h-3.5"
+              xmlns="http://www.w3.org/2000/svg"
+              width="24"
+              height="24"
+              viewBox="0 0 24 24"
+              fill="none"
+              stroke="currentColor"
+              stroke-width="2"
+              stroke-linecap="round"
+              stroke-linejoin="round"
+              ><path d="M1 12s4-8 11-8 11 8 11 8-4 8-11 8-11-8-11-8z" /><circle
+                cx="12"
+                cy="12"
+                r="3"
+              /></svg
+            >
```
Sorry this wasn't clear but this icon is in feather icons, you can import it from `svelte-feather-icons` like the other icons in the project!

ekzhang commented on src/lib/srocket.ts:
```diff
@@ -62,11 +62,14 @@ export class Srocket<T, U> {
     const data = encode(message);
+
+    // Convert buffer to Uint8Array
+    const uint8Data = new Uint8Array(data);
```
(curious why this is needed now?)

ChetanXpro replied:
the conversion is actually needed because encode() returns a node.js Buffer, but our buffer array and WebSocket interface expect a Uint8Array. Without the conversion, typeScript gives an error since Buffer and Uint8Array are different types (even though they're similar under the hood). The error specifically mentions that Buffer isn't assignable to Uint8Array<ArrayBufferLike>

ekzhang replied:
That's strange, it shouldn't be returning a Node.js buffer. It's running in the browser. I think the type checking configuration might be a little borked

ekzhang replied:
Filed a ticket https://github.com/kriszyp/cbor-x/issues/120

ekzhang reviewed (approved):
Very good! Only a couple minor things and adding a test. Feel free to make the changes if you'd like, otherwise I'm happy to take the rest (will be just a few minutes and then the test).

Thanks so much for your help! Happy new year.

ChetanXpro commented:
fixed most of the issues. Let me know if I missed something.
also, I added a test for read/write permissions. Should we cover more in the test, or is this enough?

ekzhang commented:
Test looks great 👍 — just what I was looking for!

lukehsiao commented:
Thank you for this! I'm super excited for this feature, I suspect this will push sshx to be the default for terminal sharing at $WORK.
